service/src/main/java/org/databiosphere/workspacedataservice/dataimport/ImportDetails.java

```
@@ -1,31 +1,26 @@
 package org.databiosphere.workspacedataservice.dataimport;
 
-import jakarta.annotation.Nullable;
 import java.util.UUID;
+import java.util.function.Supplier;
 import org.databiosphere.workspacedataservice.recordsink.RawlsAttributePrefixer.PrefixStrategy;
 
 public record ImportDetails(
-    @Nullable UUID jobId,
-    @Nullable String userEmail,
+    UUID jobId,
+    Supplier<String> userEmailSupplier,
     UUID collectionId,
     PrefixStrategy prefixStrategy) {
 
-  /**
-   * Convenience constructor that sets jobId and userEmail to null
-   *
-   * @param collectionId target collection
-   * @param prefixStrategy strategy for potentially renaming attributes
-   */
-  public ImportDetails(UUID collectionId, PrefixStrategy prefixStrategy) {
-    this(null, null, collectionId, prefixStrategy);
-  }
+  private static final Supplier<String> DEFAULT_EMAIL_SUPPLIER = () -> "unknown";
+  private static final UUID DEFAULT_JOB_ID =
+      UUID.fromString("00000000-0000-0000-0000-000000000000");
 
   /**
-   * Convenience constructor that sets jobId and userEmail to null, and sets prefixStrategy to NONE
+   * Convenience constructor that sets jobId and userEmail to defaults, and sets prefixStrategy to
+   * NONE. Used by TSV and JSON APIs, which are not async and thus do not have job ids
    *
    * @param collectionId target collection
    */
   public ImportDetails(UUID collectionId) {
-    this(null, null, collectionId, PrefixStrategy.NONE);
+    this(DEFAULT_JOB_ID, DEFAULT_EMAIL_SUPPLIER, collectionId, PrefixStrategy.NONE);
   }
 }
```

Use of a convenience constructor from `TdrManifestQuartzJob` was the root cause of snapshot import failing in the control plane - it set jobId to null, but `RawlsRecordSink` requires a non-null jobid.

I've removed one of the two convenience constructors, which was only used in tests, and changed the class so everything is always nonnull. One could make an argument that using default values is ugly and `null` is more appropriate; I could be convinced to switch that back.

---

> One could make an argument that using default values is ugly and null is more appropriate; I could be convinced to switch that back.

I think I would argue that.

Or... if the convenience constructor is only there tests, maybe a convenience constructor isn't the best way to go about it and we could have a test utility that creates these objects?

---

I think I prefer explicit `@Nullable` (and then making downstream classes deal with this accordingly) over having defaults, which could very easily delay or hide failure modes.

Ultimately, I'd _really_ prefer that `ImportDetails` was a more cohesive set of attributes that were always all set, but I realize that's a yakshave for another day.

---

Squinting more at this...now that TDR _does_ set the JobId, why not make it non-nullable and require it to always be set (and not let it be bypassed by the convenience constructor)?  I'd be okay with leaving in the DEFAULT_EMAIL_SUPPLIER as is.

---

I removed the convenience constructor that was only used by tests. The remaining convenience constructor _is_ used and is necessary for TSV import and JSON-api batchUpsert - both of those are synchronous and thus don't generate a jobId. I will look at putting nulls back in here and what impact that has.

---

ok, I've reverted these nullability changes and we're back to allowing jobId and userEmailSupplier to be null. I've added fail-fast validation into `RawlsRecordSinkFactory` to check that these values are non-null before even creating a `RecordSink`. How does this look?

---

The fact that you ran into this was bugging me so I tried to figure out how to `un-@Nullable` `ImportDetails` in a PR off this PR: https://github.com/DataBiosphere/terra-workspace-data-service/pull/681